oeps/oep-0038-Data-Modeling.rst

```diff
@@ -0,0 +1,419 @@
+
+- For example if you are adding a boolean to flag that a learner has not yet activated their account, the default value should be set to False, not None.
+
+The database layer should preserve uniqueness
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- If a model needs to preserve uniqueness between many fields use :code:`unique_together`_.
+.. _unique_together: https://docs.djangoproject.com/en/2.2/ref/models/options/#unique-together
+
+
+
+Best Practices
+--------------
+These practices are designed to help teams create rich and efficient data models within the OpenedX ecosystem.
+They are not standards but guidelines to help teams think about how they are storing data.
+
+Columns should store only exactly what the name describes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- The name of a column in a table should accurately describe the data in that table.
+
+- If a column is named course_id it should only store the course_id. Not the course_key, not a numeric value, not a timestamp. Just the course_id. Conversely if a column is named course_run_key it should store the course run key, not the course_id.
+
+
+Deleting data
+~~~~~~~~~~~~~
+
+-  It is better to have a column to mark the record as inactive than to
+      remove the data from the system using the SQL delete keyword. These models should use Django's
+      :code:`SoftDeletableModel`_.
+.. _SoftDeletableModel: https://django-model-utils.readthedocs.io/en/latest/models.html#softdeletablemodel
+
+
+-  Please note that GDPR may require that data be deleted. If a field
+         is determined to contain PII and falls under the realm of GDPR,
+         That data should be deleted from the system, or obfuscated from the system. `For more information about GDPR and how to delete user data from edx please refer to this documentation`_.
+.. _For more information about GDPR and how to delete user data from edx please refer to this documentation: https://openedx.atlassian.net/wiki/spaces/PLAT/pages/930021733/User+Retirement+Tutorial+for+Developers
+
+
+Don’t trap the data
+~~~~~~~~~~~~~~~~~~~
+
+-  Each piece of information should have its own column. Avoid storing
+      data in blob fields or as JSON in the database.
+
+- Another example is a concatenated string with a separator. It is best to treat these data items in 2 distinct fields.
+
+-  Don’t store encoded (pickle, json, other) objects in the database! If
+      you need to run the python environment to decode the data, analyists who use SQL will have a difficult time querying and decoding this data.
+
+Store everything
+~~~~~~~~~~~~~~~~
+
+-  Storage is cheap!
+
+-  If you are unsure whether you should store something in the database or add history the answer is almost always yes. Store the data and add history. It can be removed later if it is found unnecessary.
+
+-  Still not sure? The default answer is yes.
```

Comment:
Once you choose to persist certain data, it can be used by others whether or not the data or its usage is intentionally thought through, which then makes it harder to remove later.  We find this issue at the application layer, with other APIs, for example.  Although theoretically, one would think we should default to adding APIs, without deliberate consideration of the APIs, we will (1) be cornered into supporting them (2) remove them and break our stakeholders, or (3) spend valiant efforts to migrate them.

Given that this OEP takes the standpoint of supporting stakeholders with direct access to our data models, our data models become an API that we must support.  Moving forward, it's best for our APIs to be as intentional as possible.  Alternatively, we can annotate/designate which data models are supported APIs and which aren't (for example, in experimental/prototype phase).

Reply:
I don't feel that we should be overly restrictive about data or our data models in a broad sense. While the removal or fields has happened and caused issues in the past once or twice I don't want us to overly index on this. The removed fields were not business critical. 

That said I think that we could identify a subset of models that are perhaps more important and should have an API put in place. That said I'm unsure that should fall under the scope of this OEP. At minimum I would prefer to get this merged and deal with this for the first revision.

Reply:
As arbiter, I agree.  Let's get this version of this OEP out into the real world, and see what happens when people actually use it.  :-)